tests/unit/test_contracts/array_tests.cpp

```diff
@@ -0,0 +1,192 @@
+#include <eosio/eosio.hpp>
+#include <eosio/transaction.hpp>
+#include <eosio/table.hpp>
+
+using namespace eosio;
+
+class [[eosio::contract]] array_tests : public contract {
+   public:
+      using contract::contract;
+
+   TABLE tests {
+      uint64_t                    id;
+      std::array<uint8_t,32>      str;
+      uint64_t primary_key()      const { return id; }
+   };
+  
+   struct info {
+      int age;
+      std::string name;
+   };
+
+   typedef multi_index<name("tests"), tests> tests_table;
+   
+   struct my_struct {
+      uint32_t id;
+      std::array<std::array<std::string,4>,2> aastr;
```

Comment:
Better use typedef here for above array of arrays like typedef std::array<std::string,4> ary_str4 and then use array<ary_str4,2> as the type of above aastr. 
Because so far any abi-type ending with [][] would result in failure when tring to use cleos to put data into nodeos, e.g if using vector<vector<uint16_t> > vv, then using cleos to set vv as shown at https://github.com/EOSIO/nested-container/blob/main/nestcontn2a/nestcontn2a.cpp#L392 would result in  **Error 3015004: The type defined in the ABI is invalid, Error Details: uint16[][]**, the problem currently can be resolved by typedef vector<uint16_t> vec_uint16, and then use vector<vec_uint16> vv, the corrsponding type of vv would be vec_uint16[], instead of uint16[][]

Reply:
changed, use typedef std::array<std::string,4> array_string_4